rs/ledger_suite/tests/sm-tests/src/lib.rs

```diff
@@ -3670,6 +3673,383 @@ where
+    // Additionally store all pairs in an array in sorted order in `approve_pairs`.
+    // This allows us to check if the allowances returned by the `icrc103_get_allowances`
+    // endpoint are correct - they will always form a contiguous subarray of `approve_pairs`.
+    let mut approve_pairs = vec![];
+    for approver in &approvers {
+        for spender in &spenders {
+            let approve_args = ApproveArgs {
+                from_subaccount: approver.subaccount,
+                spender: *spender,
+                amount: Nat::from(10u64),
+                expected_allowance: None,
+                expires_at: None,
+                fee: Some(Nat::from(FEE)),
+                memo: None,
+                created_at_time: None,
+            };
+            let _ = send_approval(&env, canister_id, approver.owner, &approve_args)
+                .expect("approval failed");
+            approve_pairs.push((approver, spender));
+        }
+    }
+    assert!(approve_pairs.is_sorted());
+
+    // Check if given allowances match the elements of `approve_pairs` starting at index `pair_index`.
+    // Additionally check that the next element in `approve_pairs` has a different `from.owner`
+    // and could not be part of the same response of `icrc103_get_allowances`.
+    let check_allowances = |allowances: Allowances, pair_idx: usize, owner: Principal| {
+        for i in 0..allowances.len() {
+            let allowance = &allowances[i];
+            let pair = approve_pairs[pair_idx + i];
+            assert_eq!(allowance.from_account, *pair.0, "incorrect from account");
+            assert_eq!(allowance.to_spender, *pair.1, "incorrect spender account");
+        }
+        let next_pair_idx = pair_idx + allowances.len();
+        if next_pair_idx < approve_pairs.len() {
+            assert_ne!(approve_pairs[next_pair_idx].0.owner, owner);
+        }
+    };
+
+    // Create an Account that is lexicographically smaller than the given Account.
+    // In the above Account generation scheme, the returned account will fall
+    // between two approvers or spenders - we only modify the second byte of
+    // the owner slice or the last byte of the subaccount slice.
+    let prev_account = |account: &Account| {
+        if account.subaccount.unwrap() == [0u8; 32] {
+            let owner = account.owner.as_slice();
+            let prev_owner = [owner[0], owner[1] - 1];
+            Account {
+                owner: Principal::from_slice(&prev_owner),
+                subaccount: account.subaccount,
+            }
+        } else {
+            let mut prev_subaccount = account.subaccount.unwrap();
+            prev_subaccount[31] -= 1;
+            Account {
+                owner: account.owner,
+                subaccount: Some(prev_subaccount),
+            }
+        }
+    };
+
+    let mut prev_from = None;
+    for (idx, approve_pair) in approve_pairs.iter().enumerate() {
```

Comment:
nit: Would this perhaps be slightly more readable?
```suggestion
    for (idx, (from, spender)) in approve_pairs.iter().enumerate() {
```